src/renderToStream.ts

```diff
@@ -70,68 +72,97 @@ async function renderToNodeStream(
   disable: boolean,
   options: {
     debug?: boolean
+    onBoundaryError?: (err: unknown) => void
     renderToReadableStream?: typeof renderToReadableStream
     renderToPipeableStream?: typeof renderToPipeableStream
   }
 ) {
-  let resolve!: () => void
-  const promise = new Promise<void>((r) => {
-    resolve = () => r()
+  let resolveShell!: () => void
+  let resolveEnd!: () => void
+  const shellReady = new Promise<void>((r) => {
+    resolveShell = () => r()
+  })
+  const streamEnd = new Promise<void>((r) => {
+    resolveEnd = () => r()
   })
   let didError = false
   let firstErr: unknown = null
+  let fatalErr: unknown = null
   const onError = (err: unknown) => {
     didError = true
-    firstErr = firstErr || err
-    resolve()
+    firstErr ??= err
+    resolveShell()
+    // hacky solution to workaround https://github.com/facebook/react/issues/24536
+    // onError() gets called first, so we need to wait until next tick
+    setTimeout(() => {
+      // filter out fatal errors
+      if (fatalErr !== err) {
+        options.onBoundaryError?.(err)
+      }
+    }, 0)
   }
   const renderToPipeableStream_ = options.renderToPipeableStream ?? renderToPipeableStream
   assertReactImport(renderToPipeableStream_, 'renderToPipeableStream')
   const { pipe: pipeOriginal } = renderToPipeableStream_(element, {
-    onAllReady() {
-      resolve()
-    },
     onShellReady() {
       if (!disable) {
-        resolve()
+        resolveShell()
```

Comment:
Shouldn't the shell always resolve? And then `if (disable) await streamEnd`.

Reply:
I think your code works, but the naming is confusing. And `if (disable) await streamEnd` aligns well with the web counterpart.

Maybe we can even move up `if (disable) await streamEnd` to `renderToStream()`.

Reply:
Made some changes for this